doc/man/man1/dscp.1
.\" Man page generated from reStructuredText.
.
.TH DSCP 1 "October 23, 2013" "" ""
.SH NAME
dscp \- Distributed secure copy
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.sp
\fBdscp\fP [options] file destination
.SH DESCRIPTION
.sp
\fBdscp\fP is a tool for copying a file from one system to many systems simultaneously. It essentially performs a \fBfor a in $(seq 1 10); do scp file $a:destination; done\fP in bourne shell.
.sp
The tool assumes that dsh groups are setup either in \fB/etc/dsh/group\fP or \fB$HOME/.dsh/group\fP and that the \fBscp\fP command is available. All options are made to closely match those of dsh(1) and scp(1) as much as possible.
.SH OPTIONS
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.TP
.B \-h\fP,\fB  \-\-help
Show this help message and exit.
.TP
.B \-v\fP,\fB  \-\-verbose
Verbose output. Does not increase verbosity of \fBscp\fP .
.TP
.B \-q\fP,\fB  \-\-quiet
Make scp output quieter.
.TP
.B \-V\fP,\fB  \-\-version
Display version information and quit.
.TP
.B \-M\fP,\fB  \-\-show\-machine\-names
Prepends machine names on the standard output.
.TP
.BI \-g \ groupname\fP,\fB \ \-\-group \ groupname
Add  all  machines found in \fB/etc/dsh/group/groupname\fP to the list of machines that the specified file is copied to.
.TP
.B \-a\fP,\fB  \-\-all
Add all machines found in \fB/etc/dsh/machines.list\fP to the list of machines that the specified file is copied to.
.TP
.B \-c\fP,\fB  \-\-concurrent
Copy files concurrently.
.TP
.BI \-i \ identity_file\fP,\fB \ \-\-identity\-file \ identity_file
Selects the file from which the identity (private key) for public key authentication is read. This option is directly passed to ssh(1).
.TP
.B \-p\fP,\fB  \-\-preserve
Preserves modification times, access times, and modes from the original file.
.TP
.B \-r\fP,\fB  \-\-recursive
Recursively copy entire directories. Note that \fBscp\fP follows symbolic links encountered in the tree traversal.
.UNINDENT
.UNINDENT
.UNINDENT
.SH LICENSE
.sp
This software is released under the MIT license.
.SH SEE ALSO
.sp
ssh(1), scp(1), dsh(1)
.SH AUTHOR
Khosrow E.
.\" Generated by docutils manpage writer.
.


dscp/dscp.py
"""
dscp - distributed scure copy.
Copy a file to nodes in a cluster of systems.
"""

from __future__ import print_function
import __init__ as appinfo
import argparse
import sys
import subprocess
import logging
import os


def main():
    try:
        groupdirs = ['/etc/dsh/group', os.path.expanduser('~') + "/.dsh/group"]
    except IOError:
        groupdirs = ['/etc/dsh/group']

    try:
        machinelist = ['/etc/dsh/machines.list', os.path.expanduser('~') + "/.dsh/machines.list"]
    except IOError:
        machinelist = ['/etc/dsh/machines.list']

    parser = argparse.ArgumentParser(description=__doc__, usage="%(prog)s [options] file destination")

    # General options
    parser.add_argument("-v", "--verbose", help="Verbose output", action="store_true")
    parser.add_argument("-V", "--version", action="version", version="%(prog)s " + appinfo.__version__ )
    parser.add_argument("-q", "--quiet", help="Make scp output quieter", action="store_true")

    # dsh options
    parser.add_argument("-M", "--show-machine-names", help="Prepend the hostname on output", action="store_true")
    parser.add_argument("-g", "--group", metavar='groupname', help="Copy to group members")
    parser.add_argument("-a", "--all", help="Exectue on all machines", action="store_true")
    parser.add_argument("-c", "--concurrent", help="Copy files concurrently", action="store_true")

    # scp options
    parser.add_argument("-i", "--identity-file", metavar="identity_file", help="File containing the private key")
    parser.add_argument("-p", "--preserve", help="Preserve modification times", action="store_true")
    parser.add_argument("-r", "--recursive", help="Recursively copy entire directories", action="store_true")

    # file arguments
    parser.add_argument("file", nargs='+')
    parser.add_argument("destination")

    args = parser.parse_args()

    logging.basicConfig(format='[%(levelname)s]: %(message)s')
    logger = logging.getLogger('dscp')
    if args.verbose:
        logger.setLevel(logging.DEBUG)

    if not args.all and not args.group:
        errormsg = "One of -g or -a is required!"
        logger.error(errormsg)
        sys.exit(1)

    if args.all:
        if os.path.exists(machinelist[0]):
            filename = machinelist[0]
        elif len(machinelist) == 2 and os.path.exists(machinelist[1]):
            filename = machinelist[1]
        else:
            errormsg = "Unable to find %s" % " or ".join(machinelist)
            logger.error(errormsg)
            sys.exit(1)
    elif args.group:
        if os.path.exists(groupdirs[0] + "/" + args.group):
            filename = groupdirs[0] + "/" + args.group
        elif len(groupdirs) == 2 and os.path.exists(groupdirs[1] + "/" + args.group):
            filename = groupdirs[1] + "/" + args.group
        else:
            errormsg = "Unable to find %s in %s" % (args.group, " or ".join(groupdirs))
            logger.error(errormsg)
            sys.exit(1)

    if args.verbose:
        logger.debug("Opening file " + filename)
    try:
        nodes = open(filename)
    except IOError as e:
        logger.error(e)
        sys.exit(1)

    # Child process are kept track with a list
    children = list()

    # Wrap the execution in try/except to catch a ctrl-c
    try:
        for node in nodes:
            remote = node.rstrip()
            
            for f in args.file:
                # prepare the scp commandline
                commands = ['scp']
                if args.quiet:
                    commands.append('-q')
                if args.preserve:
                    commands.append('-p')
                if args.identity_file:
                    commands.append('-i')
                    commands.append(args.identity_file)
                if args.recursive:
                    commands.append('-r')
                commands.append(f)
                commands.append(node.rstrip() + ":" + args.destination)
                logger.debug(" ".join(commands))

                if args.show_machine_names:
                    print("Copying %s -> %s:%s" % (f, remote, args.destination))
                    sys.stdout.flush()

                try:
                    if args.concurrent:
                        # popen will spawn a child process
                        p = subprocess.Popen(commands)
                        children.append(p)
                    else:
                        ret = subprocess.call(commands)
                        if ret:
                            logger.error("%s: scp exited with error!" % remote)
                except subprocess.CalledProcessError as e:
                    logger.error(e.output)
                except OSError as e:
                    logger.error(e)
                except ValueError as e:
                    logger.error(e)

        # now wait for all children to finish
        for p in children:
            p.wait()
    except KeyboardInterrupt:
        # kill any children forcefully then quit
        print("killing all child processes...")
        for p in children:
            p.kill()
        print("leaving abruptly!", file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("leaving abruptly!", file=sys.stderr)
        sys.exit(1)
